proposals/0000-borrow-operator.md

```diff
@@ -0,0 +1,189 @@
+* Authors: ...
+* Review Manager: TBD
+* Status: ...
+
+<!--
+
+*During the review process, add the following fields as needed:*
+
+* Implementation: [apple/swift#NNNNN](https://github.com/apple/swift/pull/NNNNN) or [apple/swift-evolution-staging#NNNNN](https://github.com/apple/swift-evolution-staging/pull/NNNNN)
+* Decision Notes: [Rationale](https://forums.swift.org/), [Additional Commentary](https://forums.swift.org/)
+* Bugs: [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN), [SR-MMMM](https://bugs.swift.org/browse/SR-MMMM)
+* Previous Revision: [1](https://github.com/apple/swift-evolution/blob/...commit-ID.../proposals/NNNN-filename.md)
+* Previous Proposal: [SE-XXXX](XXXX-filename.md)
+
+-->
+
+## Introduction
+
+In this document, we propose adding a new operator to the language, marked by
+the context sensitive keyword `borrow`, that causes the compiler to pass
+arguments using a `pass-by-borrow` convention instead of a `pass-by-value`
+convention.
+
+Swift-evolution thread: [Discussion thread topic for that proposal](https://forums.swift.org/)
+
+## Motivation
+
+Swift normally passes arguments using a `pass-by-value` convention. A
+`pass-by-value` argument convention implies that each argument is passed as a
+separate value to the callee. This has the effect that the compiler is allowed
+(and sometimes required) to introduce temporary copies of arguments into the
+program in caller callsites.
+
+As an elucidating example, consider the following code:
+
+```
+func useValueTwice(_ x: Type, _ y: Type) {}
+
+var a: Type = ...
+useValueTwice(a, a)
+```
+
+Since `useValueTwice` is using default Swift conventions, we pass `a` as two
+separate guaranteed parameters to `useValueTwice`. Inside `useValueTwice`, `x`
+and `y` both appears to be separate truly borrowed values in the sense that
+`useValueTwice` has immutable access to two different values and does not take
+ownership of those values. In order for the compiler to achieve this, it must
+necessarily copy `a` to ensure that it can pass two different values to
+`useValueTwice`. These copies are observable program behavior and are in fact a
+desired default behavior since otherwise in general cases we couldn't write code
+like the above that takes the same argument multiple times. We instead would
+necessarily require the compiler to insert an error similar to an exclusivity
+violation which is undesireable:
+
+```
+useValueTwiceWithoutPassByValue(a, a) // Error! Cannot pass a value twice!
+```
+
+In certain cases though, this `pass-by-value` behavior and the implicit copies
+it requires the compiler to insert may be undesireable due to the author needing
+strict performance guarantees that implicit copies like the above are never
+generated. For most programs this guarantee is not needed since the optimizer
+generally does a good enough job eliminating such copies. But in critical system
+code, good enough is not an acceptable constraint and a true semantic guarantee
+is needed.
```

Comment:
I agree. Let me see what I can come up with.

Reply:
I am not sure if putting in an example where the optimizer fails is useful since the optimizer can change and what may be a failure today may not be a failure in the future. I think it is good enough to say that the optimizer /does/ fail sometimes.